cpp/examples/arrow/execution_plan_documentation_examples.cc

```
@@ -0,0 +1,1160 @@
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+    plan->finished().Wait();
+    return arrow::Status::OK();
+}
+
+arrow::Status SourceUnionSinkExample() {
+    auto basic_data = MakeBasicBatches();
+
+    cp::ExecContext exec_context(arrow::default_memory_pool(),
+                                 ::arrow::internal::GetCpuThreadPool());
+
+    std::shared_ptr<cp::ExecPlan> plan =
+    cp::ExecPlan::Make(&exec_context).ValueOrDie();
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    cp::Declaration union_node{"union", cp::ExecNodeOptions{}};
+    cp::Declaration lhs{"source",
+                  cp::SourceNodeOptions{basic_data.schema,
+                                    basic_data.gen(/*parallel=*/false)}};
+    lhs.label = "lhs";
+    cp::Declaration rhs{"source",
+                    cp::SourceNodeOptions{basic_data.schema,
+                                      basic_data.gen(/*parallel=*/false)}};
+    rhs.label = "rhs";
+    union_node.inputs.emplace_back(lhs);
+    union_node.inputs.emplace_back(rhs);
+
+    cp::CountOptions options(cp::CountOptions::ONLY_VALID);
+    ARROW_ASSIGN_OR_RAISE(auto declr,
+    cp::Declaration::Sequence(
+            {
+                union_node,
+                {"aggregate", cp::AggregateNodeOptions{
+                  /*aggregates=*/{{"count", &options}},
+                  /*targets=*/{"a"},
+                  /*names=*/{"count(a)"},
+                  /*keys=*/{}}},
+                {"sink", cp::SinkNodeOptions{&sink_gen}},
+            })
+            .AddToPlan(plan.get()));
+
+    ABORT_ON_FAILURE(declr->Validate());
+
+    ABORT_ON_FAILURE(plan->Validate());
+    std::shared_ptr<arrow::RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+        arrow::schema({arrow::field("count(a)", arrow::int32())}),
+        std::move(sink_gen),
+        exec_context.memory_pool());
+
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
```

Why abort here instead of `ARROW_RETURN_NOT_OK`?

---

Yes, it suits better. Replaced.